### ISSUE TITLE
macOS JDK8 OpenJ9 crashes with LWJGL 2

### ISSUE THREAD

**Issue by Ghalid** [AdoptOpenJDK/openjdk-build#1056]
**Platform:**
OpenJ9 OpenJDK 1.8.0_212-b03
**Architecture:**
macOS x64

Opening a game (in this case, a version of Minecraft at or below 1.12.2) that uses LWJGL 2 crashes during startup with the following logs:
```
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:196)
	at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:231)
	at org.multimc.EntryPoint.listen(EntryPoint.java:143)
	at org.multimc.EntryPoint.main(EntryPoint.java:34)
Caused by: java.lang.UnsatisfiedLinkError: lwjgl (Not found in java.library.path)
	at java.lang.ClassLoader.loadLibraryWithPath(ClassLoader.java:1446)
	at java.lang.ClassLoader.loadLibraryWithClassLoader(ClassLoader.java:1416)
	at java.lang.System.loadLibrary(System.java:559)
	at org.lwjgl.Sys$1.run(Sys.java:72)
	at java.security.AccessController.doPrivileged(AccessController.java:647)
	at org.lwjgl.Sys.doLoadLibrary(Sys.java:66)
	at org.lwjgl.Sys.loadLibrary(Sys.java:96)
	at org.lwjgl.Sys.<clinit>(Sys.java:117)
	at ave.J(SourceFile:2539)
	at net.minecraft.client.main.Main.main(SourceFile:41)
	... 8 more
```

This application is able to startup with Hotspot. This issue does not seem to occur with applications that use LWJGL 3.

**Comment by karianna** [comment]
@Ghalid What is LWJGL and have you tried with a Hotspot OpenJDK 1.8.0_212-b03 JVM?

**Comment by sxa555** [comment]
LWJGL will be https://sourceforge.net/projects/java-game-lib/ (Current version is v3 at https://github.com/LWJGL/lwjgl3

I can't easily reproduce the samples supplied with that on a mac right now, but they do work on xLinux 
with the OpenJ9 release

**Comment by DanHeidinga** [comment]
@Ghalid Can you generate some diagnostic files to help us look at this further?

Running with this commandline option:
```
"-Xdump:system+java:events=systhrow,filter=java/lang/UnsatisfiedLinkError,request=exclusive+prepwalk+preempt"
```
will generate a javadump (text file) and a system core.

The option can be set in the commandline or or using the `OPENJ9_JAVA_OPTIONS` environment variable.

Getting that data is a good first step for us to see what's going wrong.

**Comment by BethGriggs** [comment]
I tried running the sample supplied with LWJGL2. It succeeded with HotSpot jdk8u212-b03 from AdoptOpenJDK but not the OpenJ9 version. It seems to be finding the library and then unloading it again. 

```
 lwjgl-2.9.3$ DYLD_PRINT_LIBRARIES=1 java -cp .:res:jar/lwjgl.jar:jar/lwjgl_test.jar:jar/lwjgl_util.jar:jar/jinput.jar -Djava.library.path=./native/macosx org.lwjgl.test.WindowCreationTest
``` 

```bash
dyld: loaded: /System/Library/PrivateFrameworks/ViewBridge.framework/Versions/A/ViewBridge
dyld: loaded: /Users/bethgriggs/Downloads/jdk8u212-b03/Contents/Home/jre/lib/libfontmanager.dylib
dyld: loaded: /Users/bethgriggs/Downloads/jdk8u212-b03/Contents/Home/jre/lib/./libfreetype.6.dylib
dyld: loaded: /usr/lib/libstdc++.6.dylib
dyld: loaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: unloaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: loaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: unloaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: loaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: unloaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: loaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: unloaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: loaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: unloaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: loaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: unloaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: loaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: unloaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: loaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: unloaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: loaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
dyld: unloaded: /Users/bethgriggs/Downloads/lwjgl-2.9.3/native/macosx//liblwjgl.dylib
Exception in thread "main" java.lang.UnsatisfiedLinkError: lwjgl (Not found in java.library.path)
	at java.lang.ClassLoader.loadLibraryWithPath(ClassLoader.java:1446)
	at java.lang.ClassLoader.loadLibraryWithClassLoader(ClassLoader.java:1416)
	at java.lang.System.loadLibrary(System.java:559)
	at org.lwjgl.Sys$1.run(Sys.java:72)
	at java.security.AccessController.doPrivileged(AccessController.java:647)
	at org.lwjgl.Sys.doLoadLibrary(Sys.java:66)
	at org.lwjgl.Sys.loadLibrary(Sys.java:96)
	at org.lwjgl.Sys.<clinit>(Sys.java:117)
	at org.lwjgl.opengl.Display.<clinit>(Display.java:135)
	at org.lwjgl.test.WindowCreationTest.initialize(WindowCreationTest.java:82)
	at org.lwjgl.test.WindowCreationTest.main(WindowCreationTest.java:286)   
```

**Comment by Ghalid** [comment]
> have you tried with a Hotspot OpenJDK 1.8.0_212-b03 JVM?

Yes, Hotspot works correctly.

> @Ghalid Can you generate some diagnostic files to help us look at this further?
> 
> Running with this commandline option:
> 
> ```
> "-Xdump:system+java:events=systhrow,filter=java/lang/UnsatisfiedLinkError,request=exclusive+prepwalk+preempt"
> ```
> will generate a javadump (text file) and a system core.
> 
> The option can be set in the commandline or or using the `OPENJ9_JAVA_OPTIONS` environment variable.
> 
> Getting that data is a good first step for us to see what's going wrong.

Where should I try to locate the diagnostic files? I've ran the game with that argument, as well as adding it to the `OPENJ9_JAVA_OPTIONS` environment variable but I can't seem to find any dump anywhere. I've looked at the game directory and my home folder, and a dump didn't appear in either.

**Comment by DanHeidinga** [comment]
The size of the system core is often related to the size of the java heap.  Larger heaps = larger core files.  On the plus side, they usually compress reasonably well which helps the upload / download

**Comment by DanHeidinga** [comment]
```
1TISIGINFO     Dump Event "catch" (00000020) Detail "java/lang/UnsatisfiedLinkError" "lwjgl (dlopen(/Applications/MultiMC.app/Contents/MacOS/instances/Vanilla 1.8.9/natives/liblwjgl.dylib, 1): Library not loaded: @rpath/libjawt.dylib   Referenced from: /Applications/MultiMC.app/Contents/MacOS/instances/Vanilla 1.8.9/natives/liblwjgl.dylib   Reason: image not found)" received 
```

This is interesting in that it implies to me that there was an issue loading awt itself: `Library not loaded: @rpath/libjawt.dylib `

**Comment by DanHeidinga** [comment]
fyi @mikezhang1234567890 - this is the issue we were talking about

**Comment by mikezhang1234567890** [comment]
We can set `DYLD_PRINT_RPATHS=1` to print the rpath expansions. I ran
```
DYLD_PRINT_RPATHS=1 DYLD_PRINT_LIBRARIES=1 ../oj9-jdk8u212-b03/Contents/Home/bin/java -cp .:res:jar/lwjgl.jar:jar/lwjgl_test.jar:jar/lwjgl_util.jar:jar/jinput.jar -Djava.library.path=./native/macosx org.lwjgl.test.WindowCreationTest
```
and got this result:
```
RPATH failed to expanding     @rpath/libjawt.dylib to: /Users/mikezhang/throwaway/oj9-jdk8u212-b03/Contents/Home/jre/lib/compressedrefs/libjawt.dylib
RPATH failed to expanding     @rpath/libjawt.dylib to: /Users/mikezhang/throwaway/oj9-jdk8u212-b03/Contents/Home/bin/./libjawt.dylib
dyld: unloaded: ./native/macosx/liblwjgl.dylib
```
when `libjawt.dylib` is in `oj9-jdk8u212-b03//Contents/Home/jre/lib/libjawt.dylib`

The Hotspot lib is in `hs-jdk8u212-b03//Contents/Home/jre/lib/libjawt.dylib`, and running the test results in:
```
RPATH failed to expanding     @rpath/libjawt.dylib to: /Users/mikezhang/throwaway/hs-jdk8u212-b03/Contents/Home/jre/lib/server/./libjawt.dylib
dyld: loaded: /Users/mikezhang/throwaway/hs-jdk8u212-b03/Contents/Home/jre/lib/server/../libjawt.dylib
RPATH successful expansion of @rpath/libjawt.dylib to: /Users/mikezhang/throwaway/hs-jdk8u212-b03/Contents/Home/jre/lib/server/../libjawt.dylib
```

We may need to correct the rpath variable in OpenJ9.

**Comment by Ghalid** [comment]
Not sure if this is still needed, but I've uploaded a system core at https://drive.google.com/file/d/1hi_UYlrUAcYeQ4-eTRyxnkDzOi1CZ389/view?usp=sharing.

**Comment by mikezhang1234567890** [comment]
The newest JDK8 OpenJ9 nightly for OSX contains the fix merged above, and from my local testing, runs the LWJGL 2 sample test fine:
Running 
```
DYLD_PRINT_RPATHS=1 DYLD_PRINT_LIBRARIES=1 ../oj9-jdk8u222-b02/Contents/Home/bin/java -cp .:res:jar/lwjgl.jar:jar/lwjgl_test.jar:jar/lwjgl_util.jar:jar/jinput.jar -Djava.library.path=/Users/mikezhang/throwaway/lwjgl-2.9.3/native/macosx org.lwjgl.test.WindowCreationTest
```
gets past the exception:
```
dyld: loaded: /Users/mikezhang/throwaway/lwjgl-2.9.3/native/macosx/liblwjgl.dylib
RPATH failed to expanding     @rpath/libjawt.dylib to: /Users/mikezhang/throwaway/oj9-jdk8u222-b02/Contents/Home/jre/lib/compressedrefs/libjawt.dylib
dyld: loaded: /Users/mikezhang/throwaway/oj9-jdk8u222-b02/Contents/Home/jre/lib/compressedrefs/../libjawt.dylib
RPATH successful expansion of @rpath/libjawt.dylib to: /Users/mikezhang/throwaway/oj9-jdk8u222-b02/Contents/Home/jre/lib/compressedrefs/../libjawt.dylib
Found 14 display modes
```

**Comment by DanHeidinga** [comment]
The fix for this will be in the OpenJ9 v0.14.2 release.
https://github.com/eclipse/openj9/pull/5730
